### PR TITLE
Genetic Tuning Improvements

### PR DESCRIPTION
**Pull request by ttheodor**
Use better candidate sampling and make the selection phase more explicit and tunable by command line flags.

### CODE REVIEW

**Comment by prigoyal** [comment]
@ttheodor can you provide some information on this? I am curious to know how sampling is changing and the effect on performance. thanks!

File: src/autotuner/genetic_search.cc [hunk 1]
```diff
@@ -72,9 +74,35 @@ void mutate(
   }
 }
 
-void normalizeVector(std::vector<double>& v) {
+double mean(std::vector<double>& v) {
```
Comment (prigoyal) [inline src/autotuner/genetic_search.cc]: can these be moved to the utils file https://github.com/facebookresearch/TensorComprehensions/blob/master/src/autotuner/utils/utils.cc

where we already have mathematical functions implemented?

File: src/autotuner/genetic_search.cc [hunk 2]
```diff
@@ -72,9 +74,35 @@ void mutate(
...
+  if (v.empty()) {
+    throw std::invalid_argument("Cannot compute the mean of an empty vector.");
+  }
   auto sum = std::accumulate(v.begin(), v.end(), 0.0);
+  return sum / v.size();
+}
+
+double stdv(std::vector<double>& v, double mean) {
```
Comment (prigoyal) [inline src/autotuner/genetic_search.cc]: can these be moved to the utils file https://github.com/facebookresearch/TensorComprehensions/blob/master/src/autotuner/utils/utils.cc

where we already have mathematical functions implemented?

File: src/autotuner/genetic_search.cc [hunk 3]
```diff
@@ -72,9 +74,35 @@ void mutate(
...
+  std::transform(v.begin(), v.end(), diffs.begin(), [mean](double val) {
+    return val - mean;
+  });
+
+  auto squareSum =
+      std::inner_product(diffs.begin(), diffs.end(), diffs.begin(), 0.0);
+  return std::sqrt(squareSum / v.size());
+}
+
+void sigmaScale(std::vector<double>& v) {
```
Comment (prigoyal) [inline src/autotuner/genetic_search.cc]: can these be moved to the utils file https://github.com/facebookresearch/TensorComprehensions/blob/master/src/autotuner/utils/utils.cc

where we already have mathematical functions implemented?

**Comment by ttheodor** [comment]
@prigoyal the first commit message includes a motivation. I'll add descriptions to the others.

File: include/tc/autotuner/genetic_search.h [hunk 4]
```diff
@@ -113,6 +115,7 @@ class GeneticSearch {
   Population population;
   TuningConfiguration lastBestConf;
   const size_t kMaxPopulationSize;
+  const size_t kMatingPoolSize;
```
Comment (prigoyal) [inline include/tc/autotuner/genetic_search.h]: is this being exposed as an option to set? like we specify pop_size etc. via flags

Comment (ttheodor) [reply include/tc/autotuner/genetic_search.h]: Right now it is not. I'll expose it in a subsequent commit.

File: src/autotuner/genetic_search.cc [hunk 5]
```diff
@@ -164,6 +193,7 @@ GeneticSearch::GeneticSearch(
     : population(),
       lastBestConf(confs[0]),
       kMaxPopulationSize(n),
+      kMatingPoolSize(n * 3),
```
Comment (prigoyal) [inline src/autotuner/genetic_search.cc]: just curious: what is the reason for number 3? can we move the number to header file as a `static constexpr int`? :)

File: include/tc/autotuner/genetic_search.h (hunk 4 above)
Comment (prigoyal) [reply include/tc/autotuner/genetic_search.h]: awesome :)

**Comment by prigoyal** [comment]
thanks @ttheodor , can you update PR summary with slight context :)

File: src/autotuner/genetic_search.cc (hunk 5 above)
Comment (ttheodor) [reply src/autotuner/genetic_search.cc]: Right now it is semi-arbitrary: it is the maximum number of candidates that can be used during breeding. 

I'll expose it as an option in a subsequent commit.

**Comment by ftynse** [comment]
@ttheodor I have a hack for you: if you first push only one commit to a branch and create a PR, github will pick up the description from that commit.  Once PR is created, pushing more commits does not change the description

**Comment by prigoyal** [comment]
@caffe2bot retest this please

**Comment by ezyang** [comment]
@caffe2bot retest this please

**Comment by ezyang** [comment]
@caffe2bot retest this please

**Comment by ttheodor** [comment]
Replaced by #453

**Pull request closed without merging**